File: new_practice_python/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#
# List[int], target: int) -> List[List[int]]:
# @lc code=start
class Solution:
    def combinationSum(self, candidates, target): 
        res = []
        path = []
        candidates.sort()
        self.helper(candidates, target, path, res)

        return res




    def helper(self, candidates, target, path, res):
        # print(path)
        if target < 0: # wrong early return
            return

        if target == 0: # correct return 
            if path not in res: # 防止重複答案
                res.append(path)
                return

        for idx in range(len(candidates)):
            ele = candidates[idx]
            # print(ele)
            self.helper(candidates[idx:], target-ele, path+[ele], res) 
            # candidates 全部搜尋
            # candidates[idx:] 往後搜尋(含自己)
            # candidates[idx+1:] 往後搜尋,比自己更大的(不含自己)
            # path+[ele] 只有這次的ele, 改為list加入path







# @lc code=end

File: new_practice_cpp/39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
// https://leetcode.com/problems/combination-sum/discuss/16496/Accepted-16ms-c%2B%2B-solution-use-backtracking-easy-understand.
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> path;
        sort(candidates.begin(), candidates.end());

        dfs(candidates, target, ans, path, 0);

        return ans;   
    }

    void dfs(vector<int> &candidates, int target, vector<vector<int>> &ans, 
             vector<int> &path, int begin){
        if(target == 0){ // target == 0
            ans.push_back(path);
            return;
        }

        for (int i=begin; i < candidates.size() && target >= candidates[i]; ++i){
            path.push_back(candidates[i]);
            dfs(candidates, target - candidates[i], ans, path, i);
            path.pop_back();
        }
    }
};


// @lc code=end

